<?xml version="1.0"?>
<launch>

  <!-- Set use_sim_time parameter when you use bag data.-->
  <param name="/use_sim_time" value="true"/>

  <arg name="map_frame_id" default="map" />
  <arg name="lidar_odom_frame_id" default="odom" />

  <!-- hdl_graph_slam_node -->
  <node name="global_pose_graph_manager" pkg="hdl_graph_slam" type="hdl_graph_slam_node"  >
    <!-- New Features -->
    <param name="create_scan_ndt"     value="false" />
    <param name="use_submap_loop"     value="true" />
    <param name="leaf_voxel_size"     value="0.5" />
    <param name="ndt_leaf_min_scale"  value="0.0001" />
    <param name="min_nr"              value="10" />
    <param name="nr_submap_keyframe"  value="4" /> <!--front(nr_submap_keyframe)-current-back(nr_submap_keyframe) -->

    <param name="distance_near_thresh" value="0.1" />
    <param name="distance_far_thresh" value="4.0" />

    <param name="num_vehicle" value="2" />
    <!-- Debug Features -->

    <!-- frame settings -->
    <param name="map_frame_id" value="$(arg map_frame_id)" />
    <param name="odom_frame_id" value="$(arg lidar_odom_frame_id)" />
    <!-- optimization params -->
    <!-- typical solvers: gn_var, gn_fix6_3, gn_var_cholmod, lm_var, lm_fix6_3, lm_var_cholmod, ... -->
    <param name="g2o_solver_type" value="lm_var_cholmod" />
    <param name="g2o_solver_num_iterations" value="512" />
    <!-- keyframe registration params -->
    <param name="max_keyframes_per_update" value="20" />
    <param name="keyframe_delta_trans" value="0.5" />       <!-- meter-Default: 1.0 -->
    <param name="keyframe_delta_angle" value="0.261799" />  <!-- Radian-Default: 2.0 / 0.523599Rad=30Deg, 0.349066Rad=20Deg / 0.261799Rad=15Deg-->
    <!-- fix first node for optimization stability -->
    <param name="fix_first_node" value="true"/>
    <param name="fix_first_node_stddev" value="0.001 0.001 0.001 0.0001 0.0001 0.0001"/>
    <param name="fix_first_node_adaptive" value="true"/>
    <!-- loop closure params -->
    <param name="distance_thresh" value="2" />  <!-- Loop Closure를 매칭할 최소 위치(node-node) -->
    <param name="accum_distance_thresh" value="10" />
    <param name="min_edge_interval" value="3" />
    <param name="fitness_score_thresh" value="0.1" /> <!-- 낮을수록 매칭 잘 된 거 -->
    <!-- scan matching params -->
    <!-- ICP, GICP, NDT, GICP_OMP, NDT_OMP, FAST_GICP(recommended), or FAST_VGICP -->
    <param name="registration_method" value="NDT_OMP" />
    <param name="reg_num_threads" value="8" />
    <param name="reg_transformation_epsilon" value="0.01"/>
    <param name="reg_maximum_iterations" value="128"/>
    <param name="reg_max_correspondence_distance" value="0.4"/> <!--원래는 0.4 -->
    <param name="reg_max_optimizer_iterations" value="20"/>
    <param name="reg_use_reciprocal_correspondences" value="false"/>
    <param name="reg_correspondence_randomness" value="20"/>
    <param name="reg_resolution" value="0.3" /> <!-- ndt resolution은 이걸로 설정 -->
    <param name="reg_nn_search_method" value="DIRECT7" />
    <!-- edge params -->

    <!-- scan matching -->
    <!-- robust kernels: NONE, Cauchy, DCS, Fair, GemanMcClure, Huber, PseudoHuber, Saturated, Tukey, Welsch -->
    <param name="odometry_edge_robust_kernel" value="Cauchy" />
    <param name="odometry_edge_robust_kernel_size" value="1.0" />
    <param name="loop_closure_edge_robust_kernel" value="Huber" />
    <param name="loop_closure_edge_robust_kernel_size" value="1.0" />
    <param name="use_const_inf_matrix" value="true" />
    <param name="const_stddev_x" value="0.01" />
    <param name="const_stddev_q" value="0.001" />
    <param name="var_gain_a" value="20.0" />
    <param name="min_stddev_x" value="0.005" />
    <param name="max_stddev_x" value="0.5" />
    <param name="min_stddev_q" value="0.1" />
    <param name="max_stddev_q" value="2.0" />
    <!-- update params -->
    <param name="graph_update_interval" value="3" />
    <param name="map_cloud_update_interval" value="5" />
    <param name="map_cloud_resolution" value="0.05" />
  </node>

  <!-- <node pkg="hdl_graph_slam" type="map2odom_publisher.py" name="map2odom_publisher" /> -->

  <!-- <node name="rviz" pkg="rviz" type="rviz" args="-d $(env HOME)/graduate_ws/src/graduation_lets_go/global_pose_graph_manager/hdl_test.rviz"/> -->
  
  
  <!-- <node name="rosbag_player" pkg="rosbag" type="play" 
    args="-q $(env HOME)/graduate_ws/keyframe_engD.bag  -clock"> </node> -->


    <!-- optional arguments -->
  <!-- <arg name="enable_robot_odometry_init_guess" default="false" />
  <arg name="robot_odom_frame_id" default="robot_odom" /> -->

</launch>
